plugins/woocommerce/src/Blocks/Domain/Services/AddressAutocomplete.php

```diff
@@ -0,0 +1,97 @@
+<?php
+declare( strict_types=1 );
+namespace Automattic\WooCommerce\Blocks\Domain\Services;
+
+/**
+ * Service class for managing address autocomplete providers.
+ */
+class AddressAutocomplete {
+	/**
+	 * Registered providers
+	 *
+	 * @var array
+	 */
+	private $providers = [];
+
+	/**
+	 * Register a new address autocomplete provider.
+	 *
+	 * @param string $id   The provider ID.
+	 * @param string $name The provider name.
+	 * @return bool True if registration was successful, false otherwise.
+	 */
+	public function register_provider( string $id, string $name ): bool {
```

Comment:
Should we match other providers in Woo and ask people to provide a class? this is what Shipping and Payments do, they extend an abstract class `WC_Payment_Gateway` and `WC_Shipping_Method`. We can also match how things are registered the same way, which is via a filter. This helps race conditions in which developers don't need to time out when they register things, they just append the class to a filter and we call that when we need it.

Reply:
I think this is fine, we do give up the error checking and duplicate provider checking though don't we, if we're just operating on the result of a filter.

Reply:
You can still do that at initiation step, not registration step, which is a fine solution, we don't want to break often and early, only break as little as possible.

Reply:
Sounds good, will probably get to that in the next PR then when we actually use it.